src/shared/logger.ts

```diff
@@ -0,0 +1,216 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as os from 'os'
+import * as path from 'path'
+import * as vscode from 'vscode'
+import * as nls from 'vscode-nls'
+import * as winston from 'winston'
+import * as Transport from 'winston-transport'
+import { extensionSettingsPrefix } from './constants'
+import { DefaultSettingsConfiguration, SettingsConfiguration } from './settingsConfiguration'
+
+const localize = nls.loadMessageBundle()
+
+const LOG_RELATIVE_PATH: string = path.join('Code', 'logs', 'aws_toolkit')
+const DEFAULT_LOG_LEVEL: string = 'info'
+const DEFAULT_LOG_NAME: string = `aws_toolkit_${makeDateString('filename')}.log`
+const DEFAULT_OUTPUT_CHANNEL: vscode.OutputChannel = vscode.window.createOutputChannel('AWS Toolkit Logs')
+
+let defaultLogger: Logger
+
+export interface Logger {
+    logPath?: string
+    outputChannel?: vscode.OutputChannel
+    level: string
+    debug(...message: ErrorOrString[]): void
+    verbose(...message: ErrorOrString[]): void
+    info(...message: ErrorOrString[]): void
+    warn(...message: ErrorOrString[]): void
+    error(...message: ErrorOrString[]): void
+    releaseLogger(): void
+}
+
+export interface LoggerParams {
+    outputChannel?: vscode.OutputChannel
+    logPath?: string
+    logLevel?: string
+}
+
+// creates default logger object
+// can overwrite defaultLogger if you include params
+export function initialize(params?: LoggerParams): Logger {
+    if (!params) {
+        defaultLogger = createLogger({
+            outputChannel: DEFAULT_OUTPUT_CHANNEL,
+            logPath: getDefaultLogPath()
+        })
+        // only the default logger (with default params) gets a registered command
+        vscode.commands.registerCommand(
+            'aws.viewLogs',
+            async () => await openLogFile()
+        )
+    } else {
+        defaultLogger = createLogger(params)
+    }
+    if (defaultLogger.outputChannel) {
+        defaultLogger.outputChannel.appendLine(localize(
+            'AWS.log.fileLocation',
+            'Error logs for this session are permanently stored in {0}',
+            defaultLogger.logPath
+        ))
+    }
+
+    return defaultLogger
+}
+
+// returns default logger object
+export function getLogger(): Logger {
+    if (defaultLogger) {
+        return defaultLogger
+    }
+    throw new Error ('Default Logger not initialized. Call logger.initialize() first.')
+}
+
+// creates and returns custom logger
+// it's the caller's responsibility to keep track of this logger object
```

Comment:
If it is the caller's responsibility to call some sort of dispose/cleanup method, please add what the method(s) are in the comment.

Reply:
Do we have any conventions for adding method usage instructions to the tooltips? Might be a good place to add that there.

Reply:
In VS Code if you place your cursor on the (blank) line above a function and type `/**` and press enter, you should get the comment template. Within a comment, you can type `@` to bring up some of the sections, like `description`, `param`, and `returns`.

Reply:
https://code.visualstudio.com/docs/languages/javascript#_js-doc-support